db/system_keyspace.cc

```diff
@@ -2737,6 +2738,265 @@ class clients_table : public streaming_virtual_table {
+        bool _partition_start_emitted = false;
+        bool _partition_end_emitted = false;
+
+    private:
+        std::pair<schema_ptr, dht::partition_range> get_underlying_table_partition(const dht::decorated_key& dk) {
+            auto exploded_pk = dk.key().explode(*_schema);
+
+            const auto keyspace_name = value_cast<sstring>(utf8_type->deserialize_value(exploded_pk[0]));
+            const auto table_name = value_cast<sstring>(utf8_type->deserialize_value(exploded_pk[1]));
+            const auto underlying_pk_vals = value_cast<std::vector<data_value>>(get_underlying_key_type()->deserialize_value(exploded_pk[2]));
+
+            auto underlying_schema = _db.local().find_schema(keyspace_name, table_name);
+
+            const auto& underlying_pk_types = underlying_schema->partition_key_type()->types();
+            if (underlying_pk_types.size() != underlying_pk_vals.size()) {
+                throw std::runtime_error(fmt::format("underlying table has {} partition key components, but got {}", underlying_pk_types.size(), underlying_pk_vals.size()));
+            }
+            std::vector<bytes> underlying_pk_raw;
+            underlying_pk_raw.reserve(underlying_pk_types.size());
+            for (unsigned i = 0; i < underlying_pk_types.size(); ++i) {
+                const auto& type = underlying_pk_types[i];
+                const auto& value = underlying_pk_vals[i];
+                const auto component_str = value_cast<sstring>(value);
+                underlying_pk_raw.push_back(type->from_string(component_str));
+            }
+            auto underlying_pk = partition_key::from_exploded(underlying_pk_raw);
+            auto underlying_dk = dht::decorate_key(*underlying_schema, underlying_pk);
+
+            return std::pair(underlying_schema, dht::partition_range::make_singular(std::move(underlying_dk)));
+        }
+
+        // Runs on _shard
+        static foreign_ptr<std::unique_ptr<remote_state>>
+        create_remote_state(replica::database& db, schema_ptr output_schema, schema_ptr underlying_schema, reader_permit permit) {
+            auto& tbl = db.find_column_family(underlying_schema);
+            auto rs = std::make_unique<remote_state>(std::move(output_schema), std::move(underlying_schema), std::move(permit));
+
+            // Data sources have to be inserted ascending name order.
+            rs->data_sources.push_back(data_source{"memtable", mutation_source([&tbl] (schema_ptr schema, reader_permit permit, const dht::partition_range& pr, const query::partition_slice& ps) {
+                return tbl.make_memtable_reader(std::move(schema), std::move(permit), pr, ps);
+            })});
+            rs->data_sources.push_back(data_source{"row_cache", mutation_source([] (schema_ptr schema, reader_permit permit, const dht::partition_range& pr, const query::partition_slice& ps) {
+                //TODO
+                return make_empty_flat_reader_v2(std::move(schema), std::move(permit));
```

Comment:
This would need to be a special row_cache reader that doesn't fall back to sstables on misses.

Reply:
Yes, turns out I can reuse the memtable's existing reader for this (done in v2).